src/model.jl

```diff
@@ -58,18 +58,48 @@ Given a Case object and a DemandFlexibility object, build a matrix of demand fle
     by (bus, hour) for this interval.
 """
 function _make_bus_demand_flexibility_amount(
-    case::Case, demand_flexibility::DemandFlexibility, start_index::Int, end_index::Int
+    case::Case,
+    demand_flexibility::DemandFlexibility,
+    start_index::Int,
+    end_index::Int,
+    bus_demand::Matrix,
+    sets::Sets,
 )::Tuple{Matrix,Matrix}
-    # Bus weighting
-    zone_to_bus_shares = _make_bus_demand_weighting(case)
-    # Demand flexibility up profile
-    simulation_demand_flex_amt_up = Matrix(
-        demand_flexibility.flex_amt_up[start_index:end_index, 2:end]
-    )
-    # Demand flexibility down profile
-    simulation_demand_flex_amt_dn = Matrix(
-        demand_flexibility.flex_amt_dn[start_index:end_index, 2:end]
-    )
+
+    # combine two profiles with user input overriding DOE flexibility
+    simulation_demand_flex_amt_up = zeros(end_index - start_index + 1, sets.num_bus)
+    simulation_demand_flex_amt_dn = zeros(end_index - start_index + 1, sets.num_bus)
+
+    # DOE flexibility, convert to MW, DOE timestamps already aligned
+    if demand_flexibility.enable_doe_flexibility
+        doe_flex_pct = Matrix(demand_flexibility.doe_flex_amt[start_index:end_index, 2:end])
+        doe_flex_mw = doe_flex_pct .* permutedims(bus_demand[sets.doe_flexible_bus_idx, :])
+        for i in 1:length(sets.doe_flexible_bus_idx)
+            simulation_demand_flex_amt_up[:, sets.doe_flexible_bus_idx[i]] = doe_flex_mw[
+                :, i
+            ]
+            simulation_demand_flex_amt_dn[:, sets.doe_flexible_bus_idx[i]] = doe_flex_mw[
+                :, i
+            ]
+        end
+    end
+
+    # replace columns with user input numbers
+    csv_flex_amt_up = Matrix(demand_flexibility.flex_amt_up[start_index:end_index, 2:end])
+    csv_flex_amt_dn = Matrix(demand_flexibility.flex_amt_dn[start_index:end_index, 2:end])
+
+    for i in 1:length(sets.csv_flexible_bus_idx)
+        simulation_demand_flex_amt_up[:, sets.csv_flexible_bus_idx[i]] = csv_flex_amt_up[
+            :, i
+        ]
+        simulation_demand_flex_amt_dn[:, sets.csv_flexible_bus_idx[i]] = csv_flex_amt_dn[
+            :, i
+        ]
+    end
```

Comment:
Perhaps I'm missing something, but with the inclusion of the ability to use the DOE demand flexibility data, it is no longer guaranteed that `demand_flexibility.flex_amt_up` and `demand_flexibility.flex_amt_dn` will not be nothing if demand flexibility is enabled. Do we need to add a check here to see if `demand_flexibility.flex_amt_up` and `demand_flexibility.flex_amt_dn` are not nothing? Otherwise, it seems that there would be an error because `demand_flexibility.flex_amt_up` and `demand_flexibility.flex_amt_dn` would not have indices.

Reply:
sorry for waiting so long, it seems to be a minor problem though. there is also another typo that I did not know until I do a test-run for this issue. if the logics seems OK I will fix the conflicts and rebase tmr.